planner/optimizer.go

```diff
@@ -382,16 +382,24 @@ func isProjectionUnique(indexes database.Indexes, po *stream.ProjectOperator, pk
 	return true
 }
 
-// UseIndexBasedOnFilterNodeRule scans the tree for the first filter node whose condition is an
-// operator that satisfies the following criterias:
+type filterNode struct {
+	path document.Path
+	v    document.Value
+	f    *stream.FilterOperator
+}
+
+// UseIndexBasedOnFilterNodeRule scans the tree for filter nodes whose conditions are
+// operators that satisfies the following criterias:
 // - is a comparison operator
 // - one of its operands is a path expression that is indexed
 // - the other operand is a literal value or a parameter
-// If found, it will replace the input node by an indexInputNode using this index.
+//
+// If one or many are found, it will replace the input node by an indexInputNode using this index,
+// removing the now irrelevant filter nodes.
+//
 // TODO(asdine): add support for ORDER BY
+// TODO(jh): clarify ranges and cost code in composite indexes case
 func UseIndexBasedOnFilterNodeRule(s *stream.Stream, tx *database.Transaction, params []expr.Param) (*stream.Stream, error) {
```

Comment:
The new code is lengthy, but it felt easier to document and read in order to understand what's going.